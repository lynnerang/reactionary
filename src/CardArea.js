import React, { Component } from 'react';
import './CardArea.scss';
import Card from './Card.js';
import Guess from './Guess.js';


class CardArea extends Component {
  constructor(props) {
    super(props);

    this.state = {
      cards: [],
      cardsLeft: [],
      randomCard: {},
      choices: ['', '', ''],
      hasTerm: this.props.mode !== 'guess' && this.props.mode !== 'choice',
      showDialog: false
    }
  }

  componentDidMount() {
    this.getCards();
  }

  getCards = () => {
    //can i concat directly?
    const appCards = this.props.data;
    const myCards = this.props.myCards || [];
    const cards = this.props.mode === 'guess' ? appCards.concat(myCards) : appCards;
    const left = JSON.parse(localStorage.getItem(`remaining ${this.props.mode} cards`)) || cards.map(card => card.id);
    this.setState({cards: cards, cardsLeft: left}, () => this.getRandomCard());
  }

  getRandomCard = () => { 
    let randomId = this.state.cardsLeft[Math.floor(Math.random() * this.state.cardsLeft.length)];
    let randomCard = this.state.cards.find(card => card.id === randomId);
    this.setState({randomCard: randomCard});
    if (this.props.mode === 'choice') {
      this.setState({choices: this.shuffle(randomCard.fakeAnswers.concat(randomCard.term))});
    }
  }

  removeCard = () => {
    let ids = [...this.state.cardsLeft];
    ids.splice(ids.indexOf(this.state.randomCard.id), 1);
    if (!ids.length) {
      setTimeout(() => this.endGame(), 2000);
    }
    this.updateCardsLeft(ids);
  }

  updateCardsLeft = ids => {
    this.setState({cardsLeft: ids});
    localStorage.setItem(`remaining ${this.props.mode} cards`, JSON.stringify(ids));
  }

  resetGame = () => {
    this.updateCardsLeft(this.state.cards.map(card => card.id));
    this.props.updateGuessCount(this.props.mode, 0);
    this.getCards();
  }

  endGame = () => {
    //build out for both, passing mode
    this.props.updateGameCount();
    this.props.checkHighScore(this.props.mode, this.state.cards.length);
    this.setState({showDialog: true});
    this.resetGame();
  }

  closeDialog = () => this.setState({showDialog: false});

  shuffle = array => {
    var currentIndex = array.length, temporaryValue, randomIndex;
    while (0 !== currentIndex) {
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
    return array;
  }

  render() {
    const totalCards = this.state.cards.length;
    const cardsLeft = this.state.cardsLeft.length;
    let guess;
    let percent = (totalCards - cardsLeft) / totalCards * 100;
    if (percent === 100) {
      percent = 99.99;
    }
    const style = {animationDelay: `-${percent}s`};

      guess = 
       <Guess mode={this.props.mode} 
              answer={this.state.randomCard.term} 
              choices={this.state.choices}
              getRandomCard={this.getRandomCard} 
              removeCard={this.removeCard}
              updateGuessCount={this.props.updateGuessCount}
       /> 

    const dialog = this.state.showDialog && 
      <div className='popup'>
        <div className='round-end-dialog' onBlur={this.closeDialog}>
          <h2>Congratulations!</h2>
          <p>You have answered all flash cards.  The game has reset so you can replay any time you feel like brushing up!</p>
          <button type='button' onClick={this.closeDialog}>OK</button>
        </div>
      </div>;

      const card = this.state.randomCard && 
        <Card hasTerm={this.state.hasTerm} 
              canDelete={false} 
              cardData={this.state.randomCard}
        />

    return (
      <section className='card-area'>
        {dialog}
        {card}
        {guess}
        <div className='third-column'>
          <article className='progress-area'>
            <p><span>{totalCards - cardsLeft}</span> of <span>{totalCards}</span></p>
            <div className='progress-chart' style={style}></div>
          </article>
          <button type='button' className='reset-btn' onClick={this.resetGame}>Reset Game</button>
        </div>
      </section>
    )
  }
}

export default CardArea;