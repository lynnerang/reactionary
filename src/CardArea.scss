@import './variables.scss';

.round-end-dialog {
  background: white;
  padding: 30px 40px;;
  border-radius: 10px;
  max-width: 600px;
  line-height: 1.2;
  display: flex;
  flex-direction: column;

  h2 {
    color: $primary-color;
  }

  button {
    border: none;
    border-radius: 5px;
    background: $secondary-color;
    padding: 10px 12px 8px 12px;
    margin-top: 30px;
    color: white;
    font-weight: 400;
    align-self: flex-end;

    &:hover {
      background: $blue;
    }
  }
}

.card-area {
  display: flex;
  justify-content: space-between;
  margin: 0px 15px;

  article {
    flex: 1;
  }
}

.third-column {
  padding: 20px;
  align-self: flex-end;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  button {
    font-size: 16px;
    color: gray;
    text-decoration: underline;
    border: none;

    &:hover {
      color: rgb(177, 52, 52);
    }
  }
}

.progress-area {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 160px;
  height: 200px;
  color: gray;

  p {
    width: 125px;
    z-index: 2;
    text-align: center;
  }

  span {
    font-weight: 600;
    color: $primary-color;
  }
}

.progress-chart {
  width: 150px;
  height: 150px;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: auto;
  position: absolute;
  background: rgb(230, 230, 230) linear-gradient(to right, rgb(230, 230, 230) 50%, $secondary-color 50%);
  color: $secondary-color;
  border-radius: 50%;
  
  
  @keyframes spin {
    to {
      transform: rotate(180deg);
    }
  }
  
  @keyframes background {
    50% {
      background-color: currentColor;
    }
  }
  
  &::after {
    content: '';
    position: absolute;
    width: 80%;
    height: 80%;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    margin: auto;
    background: white;
    border-radius: 50%;
  }

  &::before {
    content: '';
    position: absolute;
    display: block;
    background-color: inherit;
    height: 100%;
    width: 50%;
    bottom: 0;
    right: 0;
    border-radius: 0 100% 100% 0 / 50%;
    transform: rotate(0);
    transform-origin: 0 50%;
    animation: 50s spin infinite linear,
               100s background infinite step-end;
    animation-play-state: paused;
    animation-delay: inherit;
  }
}

@media only screen and (max-width:1215px) { 
 .card-area {
   flex-wrap: wrap;
 }
}

@media only screen and (max-width:725px) {
  .card-area {
    flex-direction: column;
    justify-content: center;
    align-items: center;
    text-align: center;
  }

  .guess-area,
  .card,
  .progress.area {
    display: block;
    margin: 30px 0!important;
  }

  .progress-area {
    position: relative;
    width: 100%;
    bottom: 10px;
    margin: auto;
    margin-left: 35px;
  }
}

