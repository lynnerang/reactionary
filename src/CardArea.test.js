import React from 'react';
import CardArea from './CardArea.js';
import { shallow } from 'enzyme';

const mockData = 
  [
    {
    id: 1,
    term: "Single-page Application",
    fakeAnswers: [
    "Webpack",
    "Responsive Website"
    ],
    definition: "An application that loads a single HTML page and all the necessary assets (such as JavaScript and CSS) required for the app to run. Any interactions don't require a trip to server/page reload.",
    example: "./images/example{exampleId}.png"
    },
    {
    id: 2,
    term: "Compilers",
    fakeAnswers: [
    "Bundlers",
    "Package Managers"
    ],
    definition: "These things take JavaScript code, transform it, and return it as JavaScript code in a different format.",
    example: "Babel is a popular example, and is most commonly used for React and transforms ES6 syntax into syntax that older browsers can interpret."
    }
  ];


describe('CardArea', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow (
      <CardArea mode='Guess' 
      data={mockData}
      myCards={mockData}/>
      )
  });

  afterEach(() => {
    wrapper = null;
  })

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has four default state properties', () => {
    expect(wrapper.state()).toEqual({
      cards: [],
      remainingCardIds: [],
      randomCard: {},
      hasTerm: false,
      showDialog: false
    });
  });
});