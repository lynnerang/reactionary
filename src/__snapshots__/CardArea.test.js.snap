// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardArea Should match the snapshot 1`] = `
<section
  className="card-area"
>
  <Card
    canDelete={false}
    cardData={
      Object {
        "definition": "An application that loads a single HTML page and all the necessary assets (such as JavaScript and CSS) required for the app to run. Any interactions don't require a trip to server/page reload.",
        "example": "./images/example{exampleId}.png",
        "fakeAnswers": Array [
          "Webpack",
          "Responsive Website",
        ],
        "id": 1,
        "term": "Single-page Application",
      }
    }
    hasTerm={false}
  />
  <Guess
    answer="Single-page Application"
    choices={
      Array [
        "Webpack",
        "Responsive Website",
        "Single-page Application",
      ]
    }
    getRandomCard={[Function]}
    mode="Guess"
    removeCard={[Function]}
  />
  <article
    className="progress-area"
  >
    <p>
      <span>
        0
      </span>
       of 
      <span>
        4
      </span>
    </p>
    <div
      className="progress-chart"
      style={
        Object {
          "animationDelay": "-0s",
        }
      }
    />
  </article>
</section>
`;
